File: package/ios/KetchAtt.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(KetchAtt, NSObject)

RCT_EXTERN_METHOD(getStatusString:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(requestTrackingAuthorization:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

@end

File: package/ios/KetchAtt.swift
import AppTrackingTransparency
import Foundation
import React

@objc(KetchAtt)
class KetchAtt: NSObject {
  @objc static func requiresMainQueueSetup() -> Bool { false }

  @objc func getStatusString(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    if #available(iOS 14, *) {
      resolve(ATTrackingManager.trackingAuthorizationStatus.ketchAttString)
    } else {
      resolve("notDetermined")
    }
  }
}

@available(iOS 14, *)
private extension ATTrackingManager.AuthorizationStatus {
  var ketchAttString: String {
    switch self {
    case .notDetermined: return "notDetermined"
    case .restricted: return "restricted"
    case .denied: return "denied"
    case .authorized: return "authorized"
    @unknown default: return "unknown"
    }
  }
}
